src/core/border_router/routing_manager.cpp

```diff
@@ -373,22 +471,50 @@ bool RoutingManager::ContainsPrefix(const Ip6::Prefix &aPrefix, const Ip6::Prefi
     return ret;
 }
 
-const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void) const
+const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void)
 {
-    const Ip6::Prefix *newOnLinkPrefix = nullptr;
+    const Ip6::Prefix *newOnLinkPrefix      = nullptr;
+    const Ip6::Prefix *smallestOnLinkPrefix = nullptr;
+    uint8_t            onLinkPrefixNum      = 0;
 
-    // We don't evaluate on-link prefix if we are doing
-    // Router Discovery or we have already discovered some
-    // on-link prefixes.
+    // We don't evaluate on-link prefix if we are doing Router Solicitation.
     VerifyOrExit(!mRouterSolicitTimer.IsRunning());
-    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+
+    for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
-        otLogInfoBr("EvaluateOnLinkPrefix: there is already on-link prefix %s on interface %u",
-                    mDiscoveredOnLinkPrefix.ToString().AsCString(), mInfraIfIndex);
-        ExitNow();
+        if (!mDiscoveredPrefixes[i].mIsOnLinkPrefix)
+        {
+            continue;
+        }
+
+        ++onLinkPrefixNum;
+
+        if (smallestOnLinkPrefix == nullptr ||
+            IsPrefixSmallerThan(mDiscoveredPrefixes[i].mPrefix, *smallestOnLinkPrefix))
+        {
+            smallestOnLinkPrefix = &mDiscoveredPrefixes[i].mPrefix;
```

Comment:
One more question about the on-link prefix management
- Is it the case that when things converge to steady state, only on BR (the one with smallest prefix) would be advertising its on-link prefix in its RAs?
- If so,  if that BR is disconnected/gobe, the prefix will be gone (times out) and then all infra links need to allocate new address (based on new prefix from remaining BRs).
- This sound similar to me like the OMR prefix situation and that for fault tolerance, we want to have multiple BRs add the same prefix in net-data.
- Is this an issue? Do we need to consider this?

Thoughts?

Reply:
I think it is generally a good goal to keep IPv6 addresses as stable as possible. Any change to IPv6 address will require devices to perform an SRP update to register their new address and any peers to determine that the IPv6 address has changed and perform DNS-SD again.

Reply:
yes, I think a similar strategy for OMR prefixes can be applied to on-link prefixes.